src/modules/control_allocator/ControlAllocator.cpp

```diff
@@ -0,0 +1,501 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file ControlAllocator.cpp
+ *
+ * Control allocator.
+ *
+ * @author Julien Lecoeur <julien.lecoeur@gmail.com>
+ */
+
+#include "ControlAllocator.hpp"
+
+#include <drivers/drv_hrt.h>
+#include <circuit_breaker/circuit_breaker.h>
+#include <mathlib/math/Limits.hpp>
+#include <mathlib/math/Functions.hpp>
+
+using namespace matrix;
+using namespace time_literals;
+using math::radians;
+
+ControlAllocator::ControlAllocator() :
+	ModuleParams(nullptr),
+	WorkItem(MODULE_NAME, px4::wq_configurations::rate_ctrl),
+	_loop_perf(perf_alloc(PC_ELAPSED, MODULE_NAME": cycle"))
+{
+	_control_allocation = new ControlAllocationSimple();
+	parameters_updated();
+}
+
+ControlAllocator::~ControlAllocator()
+{
+	free(_control_allocation);
+	perf_free(_loop_perf);
+}
+
+bool
+ControlAllocator::init()
+{
+	if (!_vehicle_torque_setpoint_sub.registerCallback()) {
+		PX4_ERR("vehicle_torque_setpoint callback registration failed!");
+		return false;
+	}
+
+	if (!_vehicle_thrust_setpoint_sub.registerCallback()) {
+		PX4_ERR("vehicle_thrust_setpoint callback registration failed!");
+		return false;
+	}
+
+	return true;
+}
+
+void
+ControlAllocator::parameters_updated()
+{
+	// Allocation method
+	// Do this first: in case a new method is loaded, it will be configured below
+	update_allocation_method();
+
+	// Minimum actuator values
+	matrix::Vector<float, NUM_ACTUATORS> actuator_min;
+	actuator_min(0) = _param_ca_act0_min.get();
```

Comment:
@jlecoeur How do you envision are we going to handle setting actuator limits dynamically for VTOL?

Reply:
I am not sure that we need to change limits dynamically, do we? I think we need to be able to disable an actuator during hover or fixed wing flight. 

For this, I was envisioning what is written in the list above: some metadata (implemented in a parameter, or hardcoded) that indicates whether the actuator is used during hover, during fixed-wing mode, if it is safe to be prearmed, etc. Based on this metadata, we can disable the actuator (write 0s in the effectiveness matrix), and possibly disarm it.

Reply:
@jlecoeur Makes sense. We currently use to to enforce a smooth shut down of the motors at the end of the transition. But that should anways elsewhere. So I agree, we probably do not need this.